cpp/src/streaming/cudf/shuffler.cpp

```diff
@@ -10,37 +10,147 @@
+)
+    : ctx_(std::move(ctx)),
+      shuffler_(
+          ctx_->comm(),
+          ctx_->progress_thread(),
+          op_id,
+          total_num_partitions,
+          stream,
+          ctx_->br(),
+          [this](shuffler::PartID pid) -> void {
+              ctx_->comm()->logger().trace("inserting finished partition ", pid);
+              coro::sync_wait(insert_and_notify(mtx_, cv_, ready_pids_, pid));
+          },
+          ctx_->statistics(),
+          std::move(partition_owner)
+      ) {}
+
+bool ShufflerAsync::finished() const {
+    return shuffler_.finished();
+}
+
+void ShufflerAsync::insert(std::unordered_map<shuffler::PartID, PackedData>&& chunks) {
+    shuffler_.insert(std::move(chunks));
+}
+
+void ShufflerAsync::insert_finished(std::vector<shuffler::PartID>&& pids) {
+    shuffler_.insert_finished(std::move(pids));
+}
+
+coro::task<std::vector<PackedData>> ShufflerAsync::extract_async(shuffler::PartID pid) {
+    // Wait until the partition is finished
+    auto lock = co_await mtx_.scoped_lock();
+
+    RAPIDSMPF_EXPECTS(
+        !extracted_pids_.contains(pid),
+        "partition already extracted: " + std::to_string(pid),
+        std::out_of_range
+    );
+
+    co_await cv_.wait(lock, [this, pid]() {
+        // Note: purposefully not checking for extracted_pids_.contains(pid) here,
+        // because it would require notifying the cv every time a partition is extracted.
+        // Consequence of this is that, if pid was extracted by some other task, this task
+        // would only be notified during the shuffler.finished() check.
+        return shuffler_.finished() || ready_pids_.contains(pid);
+    });
+
+    // partition not found (may have been already extracted or shuffler was finished
+    // before the pid was inserted into ready_pids_)
+    RAPIDSMPF_EXPECTS(
+        !extracted_pids_.contains(pid) && ready_pids_.erase(pid) > 0,
+        "partition already extracted or not found: " + std::to_string(pid),
+        std::out_of_range
+    );
+    extracted_pids_.emplace(pid);
+    lock.unlock();  // no longer need the lock
+
+    auto chunks = shuffler_.extract(pid);
+    // shuffler gets marked as finished when all the partitions are extracted. So, tasks
+    // waiting on the cv should be notified.
+    if (shuffler_.finished()) {
+        co_await cv_.notify_all();
+    }
+
+    co_return std::move(chunks);
+}
+
+coro::task<ShufflerAsync::ExtractResult> ShufflerAsync::extract_any_async() {
+    // wait until at least one partition is ready for extraction
+    auto lock = co_await mtx_.scoped_lock();
+    co_await cv_.wait(lock, [this]() {
+        return shuffler_.finished() || !ready_pids_.empty();
+    });
+
+    // no partitions to extract or shuffle is already finished. Gracefully return an
+    // invalid result.
+    if (ready_pids_.empty()) {
+        lock.unlock();
+        ctx_->comm()->logger().trace("no partitions to extract");
+        co_return ExtractResult::invalid();
+    }
```

Comment:
question: We're using a sentinel to indicate an invalid extract result. Would it be more ergonomic to use either `coro::expected` (or `std::expected`) or `std::optional`? Then we don't have to have a "different" type of optional thing as a return value.

Reply:
I moved this to a `std::optional` return.